src/hotspot/share/runtime/synchronizer.cpp

```diff
@@ -129,34 +129,51 @@ size_t MonitorList::unlink_deflated(Thread* current, LogStream* ls,
                                     elapsedTimer* timer_p,
                                     GrowableArray<ObjectMonitor*>* unlinked_list) {
   size_t unlinked_count = 0;
+  size_t unlink_batch = 0;
   ObjectMonitor* prev = nullptr;
-  ObjectMonitor* head = Atomic::load_acquire(&_head);
-  ObjectMonitor* m = head;
+  ObjectMonitor* m = Atomic::load_acquire(&_head);
+
   // The in-use list head can be null during the final audit.
   while (m != nullptr) {
     if (m->is_being_async_deflated()) {
-      // Find next live ObjectMonitor.
+      // Find next live ObjectMonitor. Batch up the unlinkable monitors, so we can
+      // modify the list once per batch. The batch starts at "m".
       ObjectMonitor* next = m;
       do {
         ObjectMonitor* next_next = next->next_om();
         unlinked_count++;
         unlinked_list->append(next);
         next = next_next;
-        if (unlinked_count >= (size_t)MonitorDeflationMax) {
-          // Reached the max so bail out on the gathering loop.
+        if (unlink_batch++ >= (size_t)MonitorUnlinkBatch) {
+          // Reached the max batch, so bail out of the gathering loop.
+          unlink_batch = 0;
+          break;
+        }
+        if (prev == nullptr && Atomic::load(&_head) != m) {
```

Comment:
Oh, this is an interesting idea, it did not occur to me. Let me see if we can make it work. I think we can even go back to the `head` and check if `head` can be deflated in one last shot. The benefit of doing all this would be guaranteeing that we only update the `head` once, regardless of how many batches we do, and we would not have additional branch in the gathering loop.

Reply:
This does not work, unfortunately.

There is a hidden invariant: we need to unlink *all* deflated monitors. Otherwise, we would see deflated marker in downstream code that uses the monitor lists, which would break the protocol. (There are asserts that check this.) So if we bypass the first OM _and it turns out to be deflated_, we need to deflate that OM too. But inserts can happen meanwhile, and we would lose that initial OM somewhere in the middle of the list, which means we would need to go and search for it, at very least because we would need to figure out its prev.

Took me a while to understand this. Current code avoids this by carefully tracking `head` updates and searching for new `prev` directly. I will add more verification code in this method to capture the invariant.

Reply:
What you say makes sense. Thanks for finding this unwritten post-condition, and for asserting it below!